### PR TITLE
Set up Continuous Deployment

### PR DESCRIPTION
**Pull request by slifty**
This PR enables use of Travis to push master builds to a continuous deployment
server.  The details for the server are set through the Travis UI.

Adding this functionality involved modifying our installation script so that it runs
wildfly as a system service and also create a user that travis can ssh to for
deployment.  These changes are captured in the `rhel-install.sh` script.

In order to set up deployment to a server, the following keys must be populated:

- CD_USER: `travis` if the install script was used to set up the CD server.
- CD_SERVER: whatever the IP or hostname of the CD server is.
- CD_KEY: the contents, wrapped in `"` quotes of the `~/cd.key` file if the install script was used to set up the CD server.  e.g. `"-----BEGIN RSA KEY........etc."`

With these values set, any changes to master will be deployed to the server.

It should be noted that the CD simply deploys a new ear file.  If the server's configuration
needs to be changed, or other scripts are required (e.g. schema migrations), those are not
run automatically by travis.

### CODE REVIEW

File: scripts/deploy-server.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+./wildfly-10.1.0.Final/bin/jboss-cli.sh --connect \
+	--command="deploy ~/cms-portal-services.ear"
```
Comment (cecilia-donnelly) [inline scripts/deploy-server.sh]: Should we use the `--force` option here?

File: scripts/rhel-install.sh [hunk 2]
```diff
@@ -57,29 +57,47 @@ download_and_sha1 "http://download.jboss.org/wildfly/10.1.0.Final/wildfly-10.1.0
...
+sudo adduser --no-create-home wildfly
+sudo usermod -s /sbin/nologin wildfly
+sudo chown -R wildfly:wildfly /opt/wildfly-10.1.0.Final
+sudo chown -R wildfly:wildfly /opt/wildfly
+sudo chown -R wildfly:wildfly /var/log/wildfly
 
 ## Set up wildfly server
 cat pass.txt | awk '{print "psm "$1}'>> script.txt
-eval ./wildfly-10.1.0.Final/bin/add-user.sh "$(< script.txt)"
-rm script.txt
```
Comment (cecilia-donnelly) [inline scripts/rhel-install.sh]: Did you mean to remove this line?

File: .travis.yml [hunk 3]
```diff
@@ -27,23 +27,24 @@ script:
...
   - rm deploy_rsa
 
+# Continuous Deployment
 deploy:
   provider: script
   skip_cleanup: true
   script: sh scripts/deploy-ci.sh
   on:
-    branch: 222-continuous-deployment
+    branch: master
```
Comment (cecilia-donnelly) [inline .travis.yml]: :)

File: scripts/rhel-install.sh [hunk 4]
```diff
@@ -59,6 +59,7 @@ tar -xzf wildfly-10.1.0.Final.tar.gz
 rm wildfly-10.1.0.Final.tar.gz
 sudo mv wildfly-10.1.0.Final /opt/
 sudo ln -s /opt/wildfly-10.1.0.Final /opt/wildfly
+sudo ln -s /opt/wildfly-10.1.0.Final ~/wildfly-10.1.0.Final
```
Comment (cecilia-donnelly) [inline scripts/rhel-install.sh]: Why did you add this line?

Comment (slifty) [inline scripts/rhel-install.sh]: This way the build won't break for non CD uses of the script -- our Gradle build relies on Wildfly in that location.

File: scripts/rhel-install.sh (hunk 2 above)
Comment (slifty) [inline scripts/rhel-install.sh]: Great catch -- I did not

**Comment by slifty** [comment]
I personally ran it a few times in development and it deployed, but there's no clear way to test it since it requires master to trigger.  Unless someone has any ideas!

File: scripts/deploy-server.sh (hunk 1 above)
Comment (slifty) [inline scripts/deploy-server.sh]: Ah, I suspect the answer is yes (I had tested on a fresh server, so it had worked normally)

File: scripts/rhel-install.sh (hunk 4 above)
Comment (cecilia-donnelly) [inline scripts/rhel-install.sh]: Oh, right!  Thank you :)

File: scripts/rhel-install.sh [hunk 5]
```diff
@@ -109,9 +129,20 @@ xa-data-source add \
...
+sudo adduser travis
+sudo cp ~/pass.txt /home/travis/
+sudo chown travis:travis /home/travis/pass.txt
+sudo ssh-keygen -t rsa -N "" -f cd.key
+sudo mkdir /home/travis/.ssh
+sudo chown travis:travis /home/travis/.ssh
+sudo chmod 700 /home/travis/.ssh
+sudo echo | sudo dd of=/home/travis/.ssh/authorized_keys < cd.key.pub
+sudo chown travis:travis /home/travis/.ssh/authorized_keys
+sudo chmod 600 /home/travis/.ssh/authorized_keys
```
Comment (jasonaowen) [inline scripts/rhel-install.sh]: I'm finding the steps around SSH key generation to be a bit confusing. Why do them as not-travis, and then clean up after? Why call the ssh key `cd.key`, instead of the default? Why copy the newly generated key to `authorized_keys`?

I think most of these steps can be reduced to:

```sh
sudo -u travis ssh-keygen -t rsa -N ""
```

File: scripts/rhel-install.sh [hunk 6]
```diff
@@ -8,7 +8,7 @@ function download_and_sha1 {
 }
 
 function wait_for_jboss {
-	until `$1 -c "ls /deployment" &> /dev/null`; do
+	until `$1 --user=psm --password="$(< pass.txt)" -c "ls /deployment" &> /dev/null`; do
```
Comment (jasonaowen) [inline scripts/rhel-install.sh]: What changed, that we have to specify a WildFly username and password?

File: scripts/rhel-install.sh [hunk 7]
```diff
@@ -57,29 +57,49 @@ download_and_sha1 "http://download.jboss.org/wildfly/10.1.0.Final/wildfly-10.1.0
...
+sudo adduser --no-create-home wildfly
+sudo usermod -s /sbin/nologin wildfly
+sudo chown -R wildfly:wildfly /opt/wildfly-10.1.0.Final
+sudo chown -R wildfly:wildfly /opt/wildfly
+sudo chown -R wildfly:wildfly /var/log/wildfly
 
 ## Set up wildfly server
 cat pass.txt | awk '{print "psm "$1}'>> script.txt
-eval ./wildfly-10.1.0.Final/bin/add-user.sh "$(< script.txt)"
+eval sudo /opt/wildfly/bin/add-user.sh "$(< script.txt)"
```
Comment (jasonaowen) [inline scripts/rhel-install.sh]: Should this be run as the WildFly user? `sudo -u wildfly ...`

File: scripts/deploy-ci.sh [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+set -x
+scp -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ./psm-app/cms-portal-services/build/libs/cms-portal-services.ear $CD_USER@$CD_SERVER:~ && \
+ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no $CD_USER@$CD_SERVER 'bash -s' < ./scripts/deploy-server.sh
```
Comment (jasonaowen) [inline scripts/deploy-ci.sh]: We should check host keys, rather than ignore them... Can that be another variable in Travis?

File: scripts/deploy-ci.sh [hunk 9]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+set -x
+scp -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ./psm-app/cms-portal-services/build/libs/cms-portal-services.ear $CD_USER@$CD_SERVER:~ && \
```
Comment (jasonaowen) [inline scripts/deploy-ci.sh]: If we `set -e`, we don't need to have the `&&` at the end of an already long line.

File: .travis.yml [hunk 10]
```diff
@@ -21,7 +21,30 @@ install: true
...
+
+# Build the Continuous Deployment private key
+before_deploy:
+  - echo $CD_KEY > deploy_rsa
+  - sed -i -e 's/ /\n/g' deploy_rsa
+  - head -n -4 deploy_rsa > tmp
+  - tail -n +5 tmp > deploy_rsa
+  - rm tmp
+  - sed -i '1s/^/-----BEGIN RSA PRIVATE KEY-----\n/' deploy_rsa
+  - sed -i -e "\$a-----END RSA PRIVATE KEY-----" deploy_rsa
```
Comment (jasonaowen) [inline .travis.yml]: Now I see what you mean! How would you feel about extracting these to a script and calling that? That'd let you add some documentation about how to prepare the `$CD_KEY` variable from an existing SSH private key.

File: scripts/rhel-install.sh (hunk 6 above)
Comment (slifty) [inline scripts/rhel-install.sh]: Wildfly is now being run as a service using init.d scripts and a different config file.

File: scripts/rhel-install.sh [hunk 11]
```diff
@@ -57,29 +57,49 @@ download_and_sha1 "http://download.jboss.org/wildfly/10.1.0.Final/wildfly-10.1.0
...
+sudo echo | sudo dd of=/etc/default/wildfly.conf<<EOF
+JAVA_HOME="/usr/lib/jvm/java-1.8.0"
+JBOSS_HOME="/opt/wildfly"
+JBOSS_USER=wildfly
+JBOSS_MODE=standalone
+JBOSS_CONFIG=standalone-full.xml
+STARTUP_WAIT=60
+SHUTDOWN_WAIT=60
+JBOSS_CONSOLE_LOG="/var/log/wildfly/console.log"
+JBOSS_OPTS="-b 0.0.0.0 -bmanagement 0.0.0.0"
```
Comment (slifty) [inline scripts/rhel-install.sh]: I'm fine taking it out if we don't think it's useful.

File: scripts/rhel-install.sh (hunk 7 above)
Comment (slifty) [inline scripts/rhel-install.sh]: ((I'll give that a shot; I bet that will do the trick))

File: scripts/rhel-install.sh (hunk 5 above)
Comment (slifty) [inline scripts/rhel-install.sh]: This key is part of travis setup is the only reason -- happy to just do it as travis, but the artifact should be accessible so that it can be copypasta'd to the travis UI.

File: scripts/deploy-ci.sh (hunk 8 above)
Comment (slifty) [inline scripts/deploy-ci.sh]: Unfortunately no, not without hard coding the IP address of the server in the repo -- see https://github.com/travis-ci/travis-ci/issues/7043

File: scripts/rhel-install.sh (hunk 5 above)
Comment (slifty) [inline scripts/rhel-install.sh]: Update: so that command doesn't actually accomplish what I was doing above, which is to generate a key that a remote user (e.g. Travis) can use to login to our CD server.

The private key should not live on the travis user; rather, the public key needs to be registered as an authorized key for the that user.  This means it is created locally, then copied into the authorized_keys (which doesn't exist, so it needs to be created).

I did make some changes to be more concise in this process, so hopefully it is more legible.

**Comment by slifty** [comment]
@jasonaowen changes have been implemented -- ready for another review of the code (I will update documentation this afternoon)

File: scripts/rhel-install.sh [hunk 12]
```diff
@@ -108,10 +130,30 @@ xa-data-source add \
 	--xa-datasource-properties=ServerName=localhost,PortNumber=5432,DatabaseName=psm
 EOF
 
+## Populate the database with seed data
+cat psm/psm-app/db/jbpm.sql \
+    psm/psm-app/db/seed.sql \
+  | sudo -upostgres psql
```
Comment (jasonaowen) [inline scripts/rhel-install.sh]: Around line 36 we create a `psm` PostgreSQL user and database, and then just above this we configure a WildFly data source to talk to that database as that user. `sudo -upostgresql psql` connects as the `postgres` user to the `postgres` database - it should instead be `sudo -h localhost -U psm psm`. You'll also need to set up a [`~/.pgpass` file](https://www.postgresql.org/docs/current/static/libpq-pgpass.html), since `psql` doesn't have non-interactive password authentication, and I don't think we've set it up in this script yet.

File: scripts/rhel-install.sh [hunk 13]
```diff
@@ -108,10 +130,30 @@ xa-data-source add \
...
+
+## Create a key that Travis can use to login as the user
+ssh-keygen -t rsa -N "" -f CD_KEY.env
+sudo -u travis mkdir -m 700 /home/travis/.ssh
+sudo -u travis touch /home/travis/.ssh/authorized_keys
+sudo -u travis chmod 600 /home/travis/.ssh/authorized_keys
+sudo dd of=/home/travis/.ssh/authorized_keys < CD_KEY.env.pub
+
+## Set up the wildfly password in an expected location for CD
+sudo install -o travis -g travis ~/pass.txt /home/travis/
```
Comment (jasonaowen) [inline scripts/rhel-install.sh]: TIL `install(1)`! Neat.

File: scripts/rhel-install.sh [hunk 14]
```diff
@@ -57,29 +61,49 @@ download_and_sha1 "http://download.jboss.org/wildfly/10.1.0.Final/wildfly-10.1.0
...
+sudo usermod -s /sbin/nologin wildfly
+sudo chown -R wildfly:wildfly /opt/wildfly-10.1.0.Final
+sudo chown -R wildfly:wildfly /opt/wildfly
+sudo chown -R wildfly:wildfly /var/log/wildfly
 
-## Set up wildfly server
+## Set up wildfly as a service
 cat pass.txt | awk '{print "psm "$1}'>> script.txt
-eval ./wildfly-10.1.0.Final/bin/add-user.sh "$(< script.txt)"
+eval sudo -u wildfly /opt/wildfly/bin/add-user.sh "$(< script.txt)"
```
Comment (jasonaowen) [inline scripts/rhel-install.sh]: Why do we `eval` this? Running `add-user.sh` locally (without an eval), I got this output:

```ShellSession
$ ./add-user.sh test test
Added user 'test' to file '.../wildfly-10.1.0.Final/standalone/configuration/mgmt-users.properties'
Added user 'test' to file '.../wildfly-10.1.0.Final/domain/configuration/mgmt-users.properties'
```

File: scripts/rhel-install.sh [hunk 15]
```diff
@@ -57,29 +61,49 @@ download_and_sha1 "http://download.jboss.org/wildfly/10.1.0.Final/wildfly-10.1.0
...
+STARTUP_WAIT=60
+SHUTDOWN_WAIT=60
+JBOSS_CONSOLE_LOG="/var/log/wildfly/console.log"
+JBOSS_OPTS="-b 0.0.0.0"
+EOF
+sudo cp /opt/wildfly/docs/contrib/scripts/init.d/wildfly-init-redhat.sh /etc/init.d/wildfly
+sudo chkconfig --add wildfly
+sudo chkconfig wildfly on
+sudo mkdir -p /var/log/wildfly
+sudo useradd --no-create-home wildfly --no-create-home --home-dir /opt/wildfly --shell /sbin/nologin
```
Comment (slifty) [inline scripts/rhel-install.sh]: Yes indeed

File: scripts/rhel-install.sh (hunk 14 above)
Comment (slifty) [inline scripts/rhel-install.sh]: This was a silly mistake from copy pasta at some point -- eval was there because the script was wrapped in quotes (thus making it a single parameter) -- in fact, script.txt isn't even necessary since I can just type add-user.sh psm $(<pass.txt)

File: scripts/rhel-install.sh (hunk 12 above)
Comment (slifty) [inline scripts/rhel-install.sh]: (ha no worries; I figured it out!)

File: INSTALL.md [hunk 16]
```diff
@@ -383,3 +383,25 @@ documents.  We do not yet have a recommended size to set this too, but
 we will set a sane default in `cms.properties` when we do.  See
 [Issue 263](https://github.com/OpenTechStrategies/psm/issues/263) for
 some discussion of the problem.
+
+## Continuous Deployment
```
Comment (cecilia-donnelly) [inline INSTALL.md]: Thanks for adding this, @slifty!

File: scripts/rhel-install.sh [hunk 17]
```diff
@@ -33,6 +34,9 @@ sudo systemctl enable postgresql-9.6.service
 sudo systemctl start postgresql-9.6.service
 sudo -upostgres psql -c "create user \"psm\" with password '$(< pass.txt)'"
 sudo -upostgres psql -c "create database \"psm\" with owner \"psm\" encoding='utf8' template template0"
+echo localhost:5432:psm:psm:$(< pass.txt) > ~/.pgpass
+chmod 600 ~/.pgpass
```
Comment (cecilia-donnelly) [inline scripts/rhel-install.sh]: Nice.

**Comment by cecilia-donnelly** [comment]
@slifty, go ahead and merge when you get a chance!  (I'll do it if I see this lingering, but it's always nice to be able to do the merge yourself.)

**Comment by slifty** [comment]
Hold off on the merge if you could @cecilia-donnelly -- I need to make sure that travis variables are pointed to the real server and that that I update the real server with the latest schema.

**Comment by cecilia-donnelly** [comment]
🎆 thanks @slifty!!
